### PR TITLE
Support for scifunc block m

### PR DESCRIPTION
**Pull request by diptighosalkar**

### CODE REVIEW

File: webapp/data_structures/scifunc_block_m.js [hunk 1]
```diff
@@ -52,286 +52,171 @@ function scifunc_block_m() {
...
-            if(len == 7){
-                this.popup8value_disp = new ScilabString([this.popup8value[2].toString()],[this.popup8value[3].toString()],[this.popup8value[4].toString()],[this.popup8value[5].toString()],[this.popup8value[6].toString()],[this.popup8value[7].toString()],[this.popup8value[8].toString()])
+        var deptime_2 = deptime1.toString().trim().replace(regex_parentheses, '');
+        if(regex_semicolon_comma.test(deptime_2) || Number.isNaN(parseFloat(deptime_2))){
+            alert("Answer given for is block always active (0:no, 1:yes)\nhas invalid dimension:\nwaiting for dimension 1.");
+            throw "incorrect";
+        }
+
+        var opar = this.x.model.opar;
+        if(check_call_for_sci != 2){
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: variable not initialised.

File: webapp/data_structures/scifunc_block_m.js [hunk 2]
```diff
@@ -52,286 +52,206 @@ function scifunc_block_m() {
...
+        var in1 = MatrixInverse(i1);
+        for (var i = 0; i < in1.length; i++) {
+            if(in1[i].length == 1 || in1[i].length > 2){
+                if(i == 0){
+                    alert("Answer given for input ports sizes \nhas invalid dimension:\nwaiting for dimension -1 x 2.");
+                    throw "incorrect";
+                }else{
+                    alert("Answer given for input ports sizes \nis incorrect: Inconsistent row/column dimensions");
+                    throw "incorrect";
+                }
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: This code is incorrect. It should be divided into two parts:

```
if (in1[0].length == 1 || in1[0].length > 2) {
     alert("invalid dimension");
     throw "incorrect";
}

for (var i = 1; i < in1.length; i++) {
    if (in1[0].length != in1[i].length) {
        alert("inconsistent dimension");
        throw "incorrect";
    }
}
```

File: webapp/data_structures/scifunc_block_m.js [hunk 3]
```diff
@@ -52,286 +52,206 @@ function scifunc_block_m() {
...
+        var out = MatrixInverse(o1);
+        for (var i = 0; i < out.length; i++) {
+            if(out[i].length == 1 || out[i].length > 2){
+                if(i == 0){
+                    alert("Answer given for output ports sizes \nhas invalid dimension:\nwaiting for dimension -2 x 2");
+                    throw "incorrect";
+                }else{
+                    alert("Answer given for output ports sizes \nis incorrect: Inconsistent row/column dimensions");
+                    throw "incorrect";
+                }
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: This code is incorrect. It should be divided into two parts:

```
if (out[0].length == 1 || out[0].length > 2) {
     alert("invalid dimension");
     throw "incorrect";
}

for (var i = 1; i < out.length; i++) {
    if (out[0].length != out[i].length) {
        alert("inconsistent dimension");
        throw "incorrect";
    }
}
```

File: webapp/data_structures/scifunc_block_m.js [hunk 4]
```diff
@@ -52,286 +52,206 @@ function scifunc_block_m() {
...
-            this.x.model.rpar = new ScilabDouble()
-        }
-        if(this.xx != ""){
-            this.x.model.state = new ScilabDouble([this.xx])
-        }
-        if(this.z != ""){
-            this.x.model.dstate = new ScilabDouble([this.z])
+        var opar = [];
+        opar = this.x.model.opar;
+        if(check_call_for_sci != 2){
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: Write a comment on what this variable is, what are the possible values, and where it is initialised.

File: webapp/only_scifunc_code.js [hunk 5]
```diff
@@ -1,1062 +1,927 @@
...
-            } else if (i == 29) {
-                popuplabels[i][0].innerHTML = "y7 = sin(u1)";
-            }
+    var k = 0;
+    var in1 = inverse(i1);
+    var in_1_arry = [];
+    for (var i = 0; i < in1.length; i++) {
+        for (var j = 0; j < in1[i].length; j++) {
+            in_1_arry[k] = in1[i][j];
+            k++;
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: This is a repeated code block.

Write a function which takes an array and linearises it. Call that function from here.

var in_1_arry = make_linear(in1);

File: webapp/only_scifunc_code.js [hunk 6]
```diff
@@ -1,1062 +1,927 @@
...
-                popupinputs[i][0].id = "p2input7";
-            }
+    }
+    var k = 0;
+    var out = inverse(o1);
+    var out_1_arry = [];
+    for (var i = 0; i < out.length; i++) {
+        for (var j = 0; j < out[i].length; j++) {
+            out_1_arry[k] = out[i][j];
+            k++;
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: var out_1_arry = make_linear(out);

File: webapp/only_scifunc_code.js [hunk 7]
```diff
@@ -1,1062 +1,927 @@
...
+    }
+    var k = 0;
+    var clkin1 = inverse(ci1);
+    var clkin = [];
+    for (var i = 0; i < clkin1.length; i++) {
+        for (var j = 0; j < clkin1[i].length; j++) {
+            clkin[k] = clkin1[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: var clkin = make_linear(clkin1);

File: webapp/only_scifunc_code.js [hunk 8]
```diff
@@ -1,1062 +1,927 @@
...
+    var k = 0;
+    var clkout1 = inverse(co1);
+    var clkout = [];
+    for (var i = 0; i < clkout1.length; i++) {
+        for (var j = 0; j < clkout1[i].length; j++) {
+            clkout[k] = clkout1[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: var clkout = make_linear(clkout1);

File: webapp/only_scifunc_code.js [hunk 9]
```diff
@@ -1,1062 +1,927 @@
...
+    var k = 0;
+    var x0 = inverse(xx1);
+    var x01 = [];
+    for (var i = 0; i < x0.length; i++) {
+        for (var j = 0; j < x0[i].length; j++) {
+            x01[k] = x0[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: var x01 = make_linear(x0);

File: webapp/only_scifunc_code.js [hunk 10]
```diff
@@ -1,1062 +1,927 @@
...
+    var k = 0;
+    var z0 = inverse(z1);
+    var z01 = [];
+    for (var i = 0; i < z0.length; i++) {
+        for (var j = 0; j < z0[i].length; j++) {
+            z01[k] = z0[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: var z01 = make_linear(z0);

File: webapp/only_scifunc_code.js [hunk 11]
```diff
@@ -1,1062 +1,927 @@
...
+    var k = 0;
+    var rpar0 = inverse(rpar1);
+    var rpar01 = [];
+    for (var i = 0; i < rpar0.length; i++) {
+        for (var j = 0; j < rpar0[i].length; j++) {
+            rpar01[k] = rpar0[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]:
```
var rpar01 = make_linear(rpar0);
```

File: webapp/only_scifunc_code.js [hunk 12]
```diff
@@ -1,1062 +1,927 @@
...
+    var k = 0;
+    var auto = inverse(auto01);
+    var auto1 = [];
+    for (var i = 0; i < auto.length; i++) {
+        for (var j = 0; j < auto[i].length; j++) {
+            auto1[k] = auto[i][j];
+            k++;
+        }
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]:
```
var auto1 = make_linear(auto);
```

File: webapp/only_scifunc_code.js [hunk 13]
```diff
@@ -1,1062 +1,876 @@
...
+    var z01 = make_linear(z1);
+    var rpar01 = make_linear(rpar1);
+    var auto1 = make_linear(auto01);
+
+    var ni = (in_1_arry.length/2);
+    var no = (out_1_arry.length/2);
+    var nrp = 0;
+    if(rpar01.length != 0){
+        nrp = rpar0.length * rpar0[0].length;
+    }
```
Comment (sunilshetye) [inline webapp/only_scifunc_code.js]: Is this same as rpar01.length? In that case, just use:
```
var nrp = rpar01.length;
```

File: webapp/data_structures/scifunc_block_m.js [hunk 14]
```diff
@@ -52,286 +52,224 @@ function scifunc_block_m() {
...
-            if(this.popup4value[1] != "" && this.popup4value[0] == ""){
-                this.popup4value_disp = new ScilabString([this.popup4value[1].toString()])
-            }
-            if(this.popup4value[0] != "" && this.popup4value[1] != ""){
-                this.popup4value_disp = new ScilabString([this.popup4value[0].toString()],[this.popup4value[1].toString()])
+        var in1 = MatrixInverse(i1);
+        if(isNaN(in1[0].length)){
+            if(in1.length == 1 || in1.length > 2 ){
+                alert("Answer given for input ports sizes \nhas invalid dimension:\nwaiting for dimension -1 x 2.");
+                throw "incorrect";
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: Replace this by converting vector to matrix.
```
in1 = [ in1 ];
```

File: webapp/data_structures/scifunc_block_m.js [hunk 15]
```diff
@@ -52,286 +52,224 @@ function scifunc_block_m() {
...
-
-        this.popup6value = ((arguments[0]["popup6value"]))
-        if(this.popup6value[0] == "" && this.popup6value[1] == ""){
-            this.popup6value_disp = new ScilabString([this.popup6value[0].toString()])
+        var out = MatrixInverse(o1);
+        if(isNaN(out[0].length)){
+            if(out.length == 1 || out.length > 2 ){
+                alert("Answer given for output ports sizes \nhas invalid dimension:\nwaiting for dimension -1 x 2.");
+                throw "incorrect";
+            }
```
Comment (sunilshetye) [inline webapp/data_structures/scifunc_block_m.js]: Convert the vector to a matrix instead.
```
out = [ out ];
```

File: webapp/only_scifunc_code.js (hunk 13 above)
Comment (diptighosalkar) [reply webapp/only_scifunc_code.js]: actually this is not needed, i forgot to remove

**Pull request merged**